components/web_discovery/browser/content_scraper.cc

```diff
@@ -0,0 +1,382 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
```

Comment:
Overall my concern with this content scraper is it seems to be doing parsing of web content in the browser process, which would violate the Rule of 2: https://chromium.googlesource.com/chromium/src/+/master/docs/security/rule-of-2.md. In general C++ code in the browsre process should not process untrustworthy inputs unless they are very trivial or first passed through a sanitizer.

Reply:
The comment above the class declaration provides info on the two modes of operation:

https://github.com/brave/brave-core/blob/4ec17ed3ddf4ed1cef849e88b1e40c46c7563c45/components/web_discovery/browser/content_scraper.h#L47-L58

The first mode involves gathering contents from the current page in the renderer process. The untrusted content is processed by the sandboxed renderer process.

The second mode retrieves the untrusted content via the sandboxed network service process, and processes the HTML/DOM in the browser process using Rust.

Reply:
this sounds fine to me as the processing that does happen in C++ seems to be relatively minimal but would ike @bridiver to sign off

Reply:
there are several existing methods for extracting html content from the page, why are you introducing something new here?

Reply:
i am unaware of those existing methods, could you please provide a couple pointers?